### ISSUE TITLE
New Github workflow that provide links to test-configs but reuse contribution repo for report generation

### ISSUE THREAD

**Issue by relentless-pursuit** [checkstyle/checkstyle#15234]
New Github workflow in https://github.com/checkstyle/checkstyle that will use test-config files. So less configurations to provide by user in PR.

Triggering is by `Github, generate report for <path in repo>`.

Example: `Github, generate report for AbbreviationAsWordInName/Example1`
it will use config.xml and `list-of-projects.properties` from folder https://github.com/checkstyle/test-configs/tree/main/AbbreviationAsWordInName/Example1

Existing workflow: https://github.com/checkstyle/checkstyle/blob/master/.github/workflows/diff-report.yml

We will do copy of this workflow, but trigger will be on different phrase.

**Explanation:**

_Trigger:_
- The workflow is triggered when a comment is posted on a GitHub issue or pull request that starts with "Github, generate report for".
- Example: `Github, generate report for AnnotationLocation/Example1`

_Parse Comment Job:_
- Extracts the check name and example from the comment.
- Constructs URLs for the config.xml and list-of-projects.properties files based on the extracted information.
Retrieves information about the pull request branch and commit. Example https://github.com/checkstyle/test-configs/tree/main/AnnotationLocation/Example1

_Make Report Job:_
- Only runs if the parse comment job succeeds without errors.
- Downloads necessary files and sets up the environment.
- Generates a diff report using the specified configuration and project list.
- Uploads the generated report to an AWS S3 bucket.

_Send Message Job:_
- Runs regardless of the success or failure of previous jobs.
- Prepares a message based on the outcome of previous jobs:
- If successful, includes a link to the generated report.
- If there were errors, includes error information.

Posts the prepared message as a comment on the original GitHub issue or pull request.

All updates for improvement for user error handling will be in separate PR/issues. Polishing to be robust and user flriendly wil l be target for August.

Attention:  in August we will make this workflow able to work as original by 
`Github, generate report for configs in PR description`
It will simply use configs as it was defined in original workflow, this will be in use for complicated cases when we need different configs for master and patch.

Whole plan: https://docs.google.com/document/d/1_tuJsvtgA5ejtDajOO7ndCjrYzxZMllCPsjMkAum7Oc/edit#heading=h.6zga74ur8nor

**Comment by romani** [comment]
@relentless-pursuit , please update description extensively to let other maintainers understand how it will work

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit , please update description extensively to let other maintainers understand how it will work

Done.

**Comment by nrmancuso** [comment]
> We will do copy of this workflow, but trigger will be on different phrase.

This is a disappointing proposal, honestly. The existing workflow is over-complicated and fragile. I have no desire to start this new project with the technical debt of the old one, and this is the reason why I requested for us to improve the existing one as one of the first steps in this project.

It would be wise to consider reusable workflows to help us clearly separate the moving parts of this and enforce a reasonable "API" (this also makes testing much easier). Example: the "make report" workflow should accept parameters that look like what diff.groovy expects. 


Additionally, we need to start with a more reasonable and extensible trigger syntax. I envision something like this:
```
/report generate
check: AnnotationLocation
configs: Example1, Example2
```

Reasons why:

 - This allows us to more easily parse the details we need
 - Supply multiple configs
 - Readily add new parameters as we extend this workflow.
 - It is common in all other automated workflows I have seen (Github, slack, etc.) to just use a forward slash: `/report generate`. We should use a more common trigger pattern here to allow us to extend the trigger language (see below).
 - We can deprecate the old workflow and add parameters like `config_link:` and `projects_link:` easily

Specific items:
1. Where will the documentation for this workflow live? We have abstracted away the checkstyle-tester part of this almost completely from the user, so at this point the docs should live in this repo, IMO.
2. We should make a command like `/report help`? This allows maintainers to just make this comment and the workflow would share a link to your documentation for contributors to easily follow.
3. What concurrency limits should we place on this workflow?
4. `If there were errors, includes error information.` -> How do we plan to capture and relay this information?
5. We should separate the "make report" and "upload report" concepts? They are quite different and should have different APIs.

**Comment by nrmancuso** [comment]
I will allow @romani to approve this issue as he sees fit.

I do not support reusing the existing workflow in any way, and believe that it will be simpler/easier/less pain in the long run to do what I have recommended at https://github.com/checkstyle/checkstyle/issues/15234#issuecomment-2226938924.

It is sad that scope for this project (test-tool) has grown such that we cannot take time to make something extensible and well-designed.

**Comment by romani** [comment]
if we finish sooner with what we already planned to do in summer https://docs.google.com/document/d/1_tuJsvtgA5ejtDajOO7ndCjrYzxZMllCPsjMkAum7Oc/edit#heading=h.unijco89c4wb
, we will come back to your proposal.

I am not disagree with you, but we need to finish what is already in plan for clean up and improve.

We copy is only to avoid damaging existing while we are working on new.

**Comment by nrmancuso** [comment]
> We copy is only to avoid damaging existing while we are working on new.

This doesn’t make any sense, there is no reason that we can’t support both workflows simultaneously and develop the new one. Furthermore, the new workflow should be worked on and refined in a fork to avoid tying up CI and allowing @relentless-pursuit to merge as he sees fit

**Comment by romani** [comment]
Ok, let state it like this: I am afraid that I can miss during review some regression to existing diff workflow.

@relentless-pursuit , do all debug and testing in his repository/org. 

But I still worry that I can miss something, so after merge I will test in main repo.

**Comment by rnveach** [comment]
> The workflow is triggered when a comment is posted on a GitHub issue 

What are we running regression on in an issue if there is no code attached?

> Extracts the check name and example from the comment.

Does this only support one check and one example at a time?

> Downloads necessary files and sets up the environment.

What does this mean? `Make Report Job` seems to just be duplicating what the original GH workflow does, but doesn't really explain what we are gaining here in the new job.

==

What are the plans for the new workflow? Are we going to disable the old one, forcing us to be restricted to just this new one.

**Comment by romani** [comment]
In checkstyle repository PR there is link to test PR in other testing organization/repository. It is required as workflow in PR  can not executed, due to security concerns, it is rule of GitHub. Only after merging it is actionable. So to do testing @relentless-pursuit  created test organization.

>Does this only support one check and one example at a time

It supports all config bundles that are at repo https://github.com/checkstyle/test-configs
But yes, single config bundle at the time.

>Make Report Job seems to just be duplicating what the original GH workflow does,

In most functionality yes, but user doesn't need to define links any more. User just define path in test-configs repo. All config is at comment, not a PR. 

>Are we going to disable the old one, forcing us to be restricted to just this new one?

If all like workflow more than old, yes. But for some period of time they will exist together. Eventually we will make new workflow to support PR description configs, it is a easy, but will be in future PRs. At final stage of project, we expect to propose deprecation of old workflow, if all agreed, we will do this. If no agreement, we will figure out something.

**Comment by rnveach** [comment]
> yes, single config bundle at the time.

And single check? What is your plan if we need multiple checks?

> What are we running regression on in an issue if there is no code attached?

Not answered.

**Comment by romani** [comment]
>And single check?

Yes 

> What is your plan if we need multiple checks?

There is no plans on this at all, it should be in old way: create configs as required, post them in web (gist), trigger report as it is now.

New workflow just run configs from test-configs repo. We can manually put there required configs and it will be easy to run them. 

>What are we running regression on in an issue if there is no code attached?

Here is all testing https://github.com/checkstyle-gsoc/checkstyle/pull/2

>Github, generate report for AnnotationUseStyle/Example3

Will use https://github.com/checkstyle/test-configs/tree/main/AnnotationUseStyle/Example3 configs.

**Comment by rnveach** [comment]
> What is your plan if we need multiple checks?
> There is no plans on this at all, it should be in old way
> New workflow just run configs from test-configs repo. We can manually put there required configs and it will be easy to run them.

So then how will you remove old trigger if we can't find a way to support multiple checks?

Also please confirm, is there documentation on what is needed to be done for things like new checks or new properties? Should users be raising PR here first and then need to trigger something in test-config repo? Or is this done by admins?

**Comment by romani** [comment]
>So then how will you remove old trigger if we can't find a way to support multiple checks?

In one of next PRs we will add old way of parsing PR description (copy paste).  Gdoc does have reference what trigger phrase will be for it. Actual removal will be done only if all agreed at the end of summer. Two workflows will not conflict with each other.

>Also please confirm, is there documentation on what is needed to be done for things like new checks or new properties?

Documentation is same, just trigger phrase will be different.

>Should users be raising PR here first and then need to trigger something in test-config repo?

User just created a PR in checkstyle repo and do comment `Github, generate report for AnnotationUseStyle/Example3`, nothing else. Link to report will be posted as from original workflow.

**Comment by rnveach** [comment]
> one of next PRs we will add old way of parsing PR description 

Do we have documentation of the format required by users to use and some basic explanation of how to use it? If not, point to issue which will create it.

> Also please confirm, is there documentation on what is needed to be done for things like new checks or new properties?
Documentation is same, just trigger phrase will be different.

Documentation is not the same. New trigger only pulls in configs from "test-config" repo. If we are implementing a new check, it won't exist in that repo and someone will have to add it. Similar with new or removing old properties. We need something to point to users to tell them this how they do it, otherwise they won't be able to run regression, especially if we remove old trigger.

Also, what if we have a breaking change in a check? "test-config" will have to be updated first, but we would loose all possibility to run before breaking change in any other PR?

> User just create a PR in checkstyle repo and do comment ... nothing else. 

My comment was on new checks/properties. They can't run regression with your trigger on a check that doesn't exist in "test-config" right?

**Comment by romani** [comment]
>Do we have documentation of the format required by users to use and some basic explanation of how to use it?

https://github.com/checkstyle/test-configs/tree/main?tab=readme-ov-file#using-the-configurations

https://github.com/checkstyle/test-configs/tree/main/AbbreviationAsWordInName/Example1#example1-configs

>Documentation is not the same. New trigger only pulls in configs from "test-config" repo.

Yes, we don't plan to any enhancements for existing workflow. If user is not ok with preexisting configs, he needs to use existing/original workflow.

>We need something to point to users to tell them this how they do it, otherwise they won't be able to run regression, especially if we remove old trigger.

User just need to use old workflow, it is not removed in this PR, and most likely not be this summer.

>Also, what if we have a breaking change in a check? "test-config" will have to be updated first,

This project is for usage of extracted configs from master branch of checkstyle repo.  This project was never tried to deal with complicated case. Users need to use original workflow with very specific configuration to cover cases that you mentioned.

>My comment was on new checks/properties. They can't run regression with your trigger on a check that doesn't exist in "test-config" right?

Yes, users can not use test-configs for mew properties and Checks. We have no plans to make it this summer.

**Comment by rnveach** [comment]
Then it sounds like this project isn't that useful and may not get used due to lack of documentation. You say we can't cover complicated cases, so we will never be able to remove old trigger. We won't provide documentation on how to update repo for new checks/properties. https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#report-generation provides more indepth documentation, and you don't seem to want to create any documentation issues from my feedback. I will continue to point users to other documentation until things get better.

**Comment by romani** [comment]
>We won't provide documentation on how to update repo for new checks/properties.

We have https://github.com/checkstyle/test-configs/blob/main/CONTRIBUTING.md

Now I get it.
But here is new issue
https://github.com/checkstyle/test-configs/issues/143 , we planned to do the at the end but if you need it now, here it is.

>I will continue to point users to other documentation until things get better.

We will continue to improve, but I thought we need to do updates after we changed something. As we merge more and more , we will keep updating doc as we go, to be always up today. 
For now this workflow reuses existing configs and we documented this.

Please point where is still a gap to fill.

**Comment by rnveach** [comment]
I never needed anything now. I always gave the option to create issues multiple times. I am not going to review this anymore at this time.
